blog-post/blog-post.md

```diff
@@ -0,0 +1,634 @@
+we can avoid the manipulation of invalid states and the need of runtime validations
+using `require`. Also, parametric polymorphism (a.k.a. generics) and type classes
+open up new possibilities for the development of safer libraries for the modular
+development of smart-contracts.
+
+### Compile Time Evaluation
+
+While not yet implemented, we are planning to add a robust system for compile time evaluation. We
+expect to support at least the compile time evaluation of pure expressions that do not require
+access to memory. Some key design questions that remain open relate to the degree to which we will
+support access to memory at compile time, the need for compile time specific primitives (e.g. string
+concatenation / hashing), as well as the need for reflection.
+
+We want to make sure the needs of the community are met here. Please feel free to reach out
+The Solidity Forum, or our Matrix channel if you have concrete use cases in mind for
+this feature.
+
+We will publish more once we have a concrete design / prototype implementation.
+
+## SAIL, Desugaring and the Standard Library
+
+Desugaring is the process of translating high-level language constructs into
+a semantically equivalent form using language's more primitive features.
+Some of the Core Solidity new high-level constructs will be translated
+into a intermediate language called SAIL - Solidity Algebraic Intermediate
+Language. Anonymous functions, assignments, contracts, indexed and storage
+access are examples of high-level Core Solidity features with are translated
+into SAIL. The SAIL language is a typed programming language featuring
+generics, typeclasses, algebraic data types and pattern matchings and it
+serves as a high-level intermediate step between Core Solidity and Yul.
+
+As an example of one translation steps, let's see how assignments to
+contract state variables are translated into SAIL constructs. Consider the
+following simple contract which updates the value of state variable
+named `counter`:
+
+```
+contract Counter {
+    counter : word;
+
+    function main() -> word {
+        counter += 42;
+        return counter;
+    }
+}
+```
+
+During the compilation of the previous code piece, the assignment `counter += 42`
+is desugared into:
+
+```
+Assign.assign(Lvalue(counter_sel), Num.add(Rvalue(counter_sel), 42))
+```
+
+The addition infix operator is translated into function `add`, defined by a
+type class `Num`. Access of the state variable `counter` is done by functions
+`Lvalue` and `Rvalue`, which retrieve the storage pointer and the current
+value for the `counter` variable; and function `Assign.assign` updates
+the storage using the pointer and the value returned by the addition
+operation. The complete code for this desugaring step is as follows:
+
+```
+contract Counter {
+    function main () -> word {
+        Assign.assign(Lvalue(counter_sel), Num.add(Rvalue(counter_sel), 42));
+        return Rvalue(counter_sel);
```

Comment:
What's `counter_sel` and where it is defined? I expected this to desugar to a fixed address based on storage layout, but this looks like a variable declared elsewhere. The name is also quite cryptic (sel = selector? makes no sense).

Reply:
this example has been removed entirely. explaining `counter_sel` is non trivial. I replaced it with the abi.encode example.